openpype/hosts/maya/api/lib.py

```diff
@@ -172,6 +174,223 @@ def maintained_selection():
             cmds.select(clear=True)
 
 
+def reload_all_udim_tile_previews():
+    """Regenerate all UDIM tile preview in texture file"""
+    texture_files = cmds.ls(type="file")
+    if texture_files:
+        for texture_file in texture_files:
+            if cmds.getAttr("{}.uvTilingMode".format(texture_file)) > 0:
+                cmds.ogs(regenerateUVTilePreview=texture_file)
+    cmds.ogs(reloadTextures=True)
+
+
+@contextlib.contextmanager
+def panel_camera(panel, camera):
+    """Set modelPanel's camera during the context.
+
+    Arguments:
+        panel (str): modelPanel name.
+        camera (str): camera name.
+
+    """
+    original_camera = cmds.modelPanel(panel, query=True, camera=True)
+    try:
+        cmds.modelPanel(panel, edit=True, camera=camera)
+        yield
+    finally:
+        cmds.modelPanel(panel, edit=True, camera=original_camera)
+
+
+def render_capture_preset(preset):
+    """Capture playblast with a preset.
+
+    To generate the preset use `generate_capture_preset`.
+
+    Args:
+        preset (dict): preset options
+
+    Returns:
+        str: Output path of `capture.capture`
+    """
+
+    # Force a refresh at the start of the timeline
+    # TODO (Question): Why do we need to do this? What bug does it solve?
+    #   Is this for simulations?
+    cmds.refresh(force=True)
+    refresh_frame_int = int(cmds.playbackOptions(query=True, minTime=True))
+    cmds.currentTime(refresh_frame_int - 1, edit=True)
+    cmds.currentTime(refresh_frame_int, edit=True)
+
+    if os.environ.get("OPENPYPE_DEBUG") == "1":
+        log.debug(
+            "Using preset: {}".format(
+                json.dumps(preset, indent=4, sort_keys=True)
+            )
+        )
+
+    # not supported by `capture` so we pop it off of the preset
+    reload_textures = preset["viewport_options"].pop("reloadTextures", True)
+
+    with ExitStack() as stack:
+        stack.enter_context(maintained_time())
+        stack.enter_context(panel_camera(preset["panel"], preset["camera"]))
+        stack.enter_context(viewport_default_options(preset))
+        if preset["viewport_options"].get("textures"):
+            # Force immediate texture loading when to ensure
+            # all textures have loaded before the playblast starts
+            stack.enter_context(material_loading_mode("immediate"))
+            if reload_textures:
+                # Regenerate all UDIM tiles previews
+                reload_all_udim_tile_previews()
+        preset.pop("panel")
+        path = capture.capture(log=self.log, **preset)
```

Comment:
Let's move the panel popping out of the exit stack since it's unrelated to the exit stack itself.
```suggestion
    preset.pop("panel")

    with ExitStack() as stack:
        stack.enter_context(maintained_time())
        stack.enter_context(panel_camera(preset["panel"], preset["camera"]))
        stack.enter_context(viewport_default_options(preset))
        if preset["viewport_options"].get("textures"):
            # Force immediate texture loading when to ensure
            # all textures have loaded before the playblast starts
            stack.enter_context(material_loading_mode("immediate"))
            if reload_textures:
                # Regenerate all UDIM tiles previews
                reload_all_udim_tile_previews()
        path = capture.capture(log=self.log, **preset)
```

Reply:
But if we move the preset.pop("panel") before the` enter context` functions, it errors out the extractor as the preset options doesn't include `preset["panel"]` in line 237
Maybe we can put it into right after all the `enter_context` functions?

Reply:
You are correct - ignore my comment. It needs to be there because it needs access to it - however, popping it within the context would then ALSO break because the "reverting" logic of the context managers also try to still access it! Unless the context managers do:

`panel = preset["panel"]` before the yield
OR we pass the panel along explicitly to the function.

I believe the current code will actually still error on exiting the context manager, because after the capture (after the exit stack) it tries to still access the key e.g. here https://github.com/ynput/OpenPype/blob/5f309994c39ae8600f414da8ce1da15fec729408/openpype/hosts/maya/api/lib.py#L375-L379

Reply:
@BigRoy I agree with `panel = preset.pop("panel")` makes more sense. Already updated in the latest commit.